Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 363ffde7ef1d0188afc347731fc805ef3b086743
Number of Error(s): 14
checking for strings.h... /home/centos/smoketest/PR-13413/HPCC-Platform/system/jlib/jptree.cpp:8189:1: error: version control conflict marker in file
/home/centos/smoketest/PR-13413/HPCC-Platform/system/jlib/jptree.cpp:8208:43: error: \xe2\x80\x98yaml_write_iiostream\xe2\x80\x99 was not declared in this scope
/home/centos/smoketest/PR-13413/HPCC-Platform/system/jlib/jptree.cpp:8457:1: error: version control conflict marker in file
/home/centos/smoketest/PR-13413/HPCC-Platform/system/jlib/jptree.cpp:8457:3: error: expected primary-expression before \xe2\x80\x98==\xe2\x80\x99 token
/home/centos/smoketest/PR-13413/HPCC-Platform/system/jlib/jptree.cpp:8457:5: error: expected primary-expression before \xe2\x80\x98==\xe2\x80\x99 token
/home/centos/smoketest/PR-13413/HPCC-Platform/system/jlib/jptree.cpp:8457:7: error: expected primary-expression before \xe2\x80\x98=\xe2\x80\x99 token
/home/centos/smoketest/PR-13413/HPCC-Platform/system/include/platform.h:518:22: error: expected primary-expression before \xe2\x80\x98__attribute__\xe2\x80\x99
/home/centos/smoketest/PR-13413/HPCC-Platform/system/jlib/jptree.cpp:8462:1: error: expected \xe2\x80\x98}\xe2\x80\x99 at end of input
make[2]: *** [system/jlib/CMakeFiles/jlib.dir/build.make:427: system/jlib/CMakeFiles/jlib.dir/jptree.cpp.o] Error 1
checking if gcc static flag -static works... make[1]: *** [CMakeFiles/Makefile2:1266: system/jlib/CMakeFiles/jlib.dir/all] Error 2
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
make: *** [Makefile:163: all] Error 2

[Build error:](https://gist.github.com/f9b46265c1051720c4c128083af0f542)
Build log [Build log](https://gist.github.com/HPCCSmoketest/f9b46265c1051720c4c128083af0f542#file-result-20-03-04-19-06-00-log)

--- a/common/workunit/workunit.cpp
+++ b/common/workunit/workunit.cpp
@@ -12950,6 +12950,19 @@ extern WORKUNIT_API void addTimeStamp(IWorkUnit * wu, StatisticScopeType scopeTy
     wu->setStatistic(queryStatisticsComponentType(), queryStatisticsComponentName(), scopeType, scopestr, kind, NULL, getTimeStampNowValue(), 1, 0, StatsMergeAppend);
 }
 
+extern WORKUNIT_API double calculateThorCost(__int64 timeNs, unsigned clusterWidth)
+{
+    IPropertyTree *costs = queryCostsConfiguration();
+    if (costs)
+    {
+        double thor_master_rate = costs->getPropReal("thor/@master", 0.0);
+        double thor_slave_rate = costs->getPropReal("thor/@slave", 0.0);
+
+        double time_seconds = ((double)timeNs/1000000000);
+        return (time_seconds * thor_master_rate + time_seconds * thor_slave_rate * clusterWidth)*1E6;
+    }
+    return 0;
+}
 
 void aggregateStatistic(StatsAggregation & result, IConstWorkUnit * wu, const WuScopeFilter & filter, StatisticKind search)
 {

--- a/common/workunit/workunit.hpp
+++ b/common/workunit/workunit.hpp
@@ -1655,6 +1655,8 @@ extern WORKUNIT_API const char * getWorkunitActionStr(WUAction action);
 extern WORKUNIT_API WUAction getWorkunitAction(const char * actionStr);
 
 extern WORKUNIT_API void addTimeStamp(IWorkUnit * wu, StatisticScopeType scopeType, const char * scope, StatisticKind kind, unsigned wfid=0);
+extern WORKUNIT_API double calculateThorCost(__int64 timeNs, unsigned clusterWidth);
+
 extern WORKUNIT_API IPropertyTree * getWUGraphProgress(const char * wuid, bool readonly);
 
 class WORKUNIT_API WorkUnitErrorReceiver : implements IErrorReceiver, public CInterface

--- a/dockerfiles/thormaster/thor.xml
+++ b/dockerfiles/thormaster/thor.xml
@@ -1,3 +1,6 @@
 <Thor name="mythor" daliServers="dali" numSlaves="1" pluginsPath="/opt/HPCCSystems/plugins/" replicateAsync="false"
       watchdogEnabled="true" watchdogProgressEnabled="true" globalMemorySize="4096">
+  <cost>
+    <thor master="0.000002" slave="0.00001"/>
+  </cost>
 </Thor>

--- a/system/jlib/jptree.cpp
+++ b/system/jlib/jptree.cpp
@@ -1306,6 +1306,12 @@ void PTree::addPropInt(const char *xpath, int val)
     addPropInt64(xpath, val); // underlying type always __int64 (now)
 }
 
+double PTree::getPropReal(const char *xpath, double dft) const
+{
+    const char *val = queryProp(xpath);
+    return val?atof(val):dft;
+}
+
 bool PTree::isCompressed(const char *xpath) const
 {
     if (!xpath)
@@ -8447,4 +8453,9 @@ void saveYAML(IFileIO &ifileio, const IPropertyTree *tree, unsigned indent, unsi
 void saveYAML(IIOStream &stream, const IPropertyTree *tree, unsigned indent, unsigned flags)
 {
     toYAML(tree, stream, indent, flags);
+}
+
+jlib_decl IPropertyTree * queryCostsConfiguration()
+{
+    return queryComponentConfig().queryPropTree("costs");
 }

--- a/system/jlib/jptree.hpp
+++ b/system/jlib/jptree.hpp
@@ -92,6 +92,8 @@ interface jlib_decl IPropertyTree : extends serializable
     virtual void setPropInt64(const char *xpath, __int64 val) = 0;
     virtual void addPropInt64(const char *xpath, __int64 val) = 0;
 
+    virtual double getPropReal(const char *xpath, double dft) const = 0;
+
     virtual bool getPropBin(const char *xpath, MemoryBuffer &ret) const = 0;
     virtual void setPropBin(const char *xpath, size32_t size, const void *data) = 0;
     virtual void addPropBin(const char *xpath, size32_t size, const void *data) = 0;
@@ -291,6 +293,7 @@ inline static bool isValidXPathChr(char c)
 jlib_decl IPropertyTree * loadArgsIntoConfiguration(IPropertyTree *config, const char * * argv);
 jlib_decl IPropertyTree * loadConfiguration(const char * defaultYaml, const char * * argv, const char * componentTag, const char * envPrefix, const char * legacyFilename, IPropertyTree * (mapper)(IPropertyTree *));
 jlib_decl StringBuffer & regenerateConfig(StringBuffer &jsonText, IPropertyTree * config, const char * componentTag);
+jlib_decl IPropertyTree * queryCostsConfiguration();
 
 //The following can only be called after loadConfiguration has been called.  All components must call loadConfiguration().
 jlib_decl IPropertyTree & queryGlobalConfig();

--- a/system/jlib/jptree.ipp
+++ b/system/jlib/jptree.ipp
@@ -626,6 +626,7 @@ public:
     virtual int getPropInt(const char *xpath, int dft=0) const override;
     virtual void setPropInt(const char *xpath, int val) override;
     virtual void addPropInt(const char *xpath, int val) override;
+    virtual double getPropReal(const char *xpath, double dft) const override;
     virtual bool getPropBin(const char * xpath, MemoryBuffer &ret) const override;
     virtual void setPropBin(const char * xpath, size32_t size, const void *data) override;
     virtual void appendPropBin(const char *xpath, size32_t size, const void *data) override;

--- a/system/jlib/jstatcodes.h
+++ b/system/jlib/jstatcodes.h
@@ -92,6 +92,7 @@ enum StatisticMeasure
     SMeasureBool,                       // A boolean
     SMeasureId,                         // An Id for an element
     SMeasureFilename,                   // A filename
+    SMeasureCost,                       // Used to measure cost
     SMeasureMax,
 };
 
@@ -231,6 +232,7 @@ enum StatisticKind
     StSizeOsDiskWrite,
     StTimeBlocked,
     StCycleBlockedCycles,
+    StCostExecute,
     StMax,
 
     //For any quantity there is potentially the following variants.

--- a/system/jlib/jstats.cpp
+++ b/system/jlib/jstats.cpp
@@ -66,7 +66,7 @@ void setStatisticsComponentName(StatisticCreatorType processType, const char * p
 //--------------------------------------------------------------------------------------------------------------------
 
 // Textual forms of the different enumerations, first items are for none and all.
-static constexpr const char * const measureNames[] = { "", "all", "ns", "ts", "cnt", "sz", "cpu", "skw", "node", "ppm", "ip", "cy", "en", "txt", "bool", "id", "fname", NULL };
+static constexpr const char * const measureNames[] = { "", "all", "ns", "ts", "cnt", "sz", "cpu", "skw", "node", "ppm", "ip", "cy", "en", "txt", "bool", "id", "fname", "cost", NULL };
 static constexpr const char * const creatorTypeNames[]= { "", "all", "unknown", "hthor", "roxie", "roxie:s", "thor", "thor:m", "thor:s", "eclcc", "esp", "summary", NULL };
 static constexpr const char * const scopeTypeNames[] = { "", "all", "global", "graph", "subgraph", "activity", "allocator", "section", "compile", "dfu", "edge", "function", "workflow", "child", "unknown", nullptr };
 
@@ -432,6 +432,8 @@ StringBuffer & formatStatistic(StringBuffer & out, unsigned __int64 value, Stati
         return out.append(value);
     case SMeasureEnum:
         return out.append("Enum{").append(value).append("}"); // JCS->GH for now, should map to known enum text somehow
+    case SMeasureCost:
+        return out.appendf("$ %.06f", (double) value / 1E6);
     default:
         return out.append(value).append('?');
     }
@@ -613,6 +615,7 @@ const char * queryMeasurePrefix(StatisticMeasure measure)
     case SMeasureBool:          return "Is";
     case SMeasureId:            return "Id";
     case SMeasureFilename:      return "";
+    case SMeasureCost:          return "Cost";
     case SMeasureNone:          return nullptr;
     default:
         return "Unknown";
@@ -664,6 +667,7 @@ StatsMergeAction queryMergeMode(StatisticMeasure measure)
     case SMeasureBool:          return StatsMergeKeepNonZero;
     case SMeasureId:            return StatsMergeKeepNonZero;
     case SMeasureFilename:      return StatsMergeKeepNonZero;
+    case SMeasureCost:          return StatsMergeSum;
     default:
 #ifdef _DEBUG
         throwUnexpected();
@@ -878,7 +882,8 @@ static const StatisticMeta statsMetaData[StMax] = {
     { SIZESTAT(OsDiskRead) },
     { SIZESTAT(OsDiskWrite) },
     { TIMESTAT(Blocked) },
-    { CYCLESTAT(Blocked) }
+    { CYCLESTAT(Blocked) },
+    { STAT(Cost, Execute, SMeasureCost) },
 };
 
 

--- a/thorlcr/master/thgraphmanager.cpp
+++ b/thorlcr/master/thgraphmanager.cpp
@@ -919,7 +919,10 @@ bool CJobManager::executeGraph(IConstWorkUnit &workunit, const char *graphName, 
         updateWorkunitStat(wu, SSTgraph, graphName, StTimeElapsed, graphTimeStr, graphTimeNs, wfid);
 
         addTimeStamp(wu, SSTgraph, graphName, StWhenFinished, wfid);
-        
+        double cost = calculateThorCost(graphTimeNs, queryNodeClusterWidth());
+        if (cost)
+            wu->setStatistic(queryStatisticsComponentType(), queryStatisticsComponentName(), SSTgraph, graphScope, StCostExecute, NULL, cost, 1, 0, StatsMergeReplace);
+
         removeJob(*job);
     }
     catch (IException *e)

